controllers/application/ingress_virtual_service.go

```diff
@@ -71,8 +73,120 @@ func (r *ApplicationReconciler) reconcileIngressVirtualService(ctx context.Conte
+						RedirectCode: 302,
+					},
+				},
+			},
+		},
+	}
+
+	err := ctrlutil.SetControllerReference(application, &virtualService, r.GetScheme())
+	if err != nil {
+		r.SetControllerError(ctx, application, controllerName, err)
+		return &virtualService, err
+	}
+
+	r.SetLabelsFromApplication(ctx, &virtualService, *application)
+	util.SetCommonAnnotations(&virtualService)
+
+	return &virtualService, err
+}
+
+func (r *ApplicationReconciler) defineCommonVirtualService(ctx context.Context, application *skiperatorv1alpha1.Application) (*networkingv1beta1.VirtualService, error) {
+	virtualService := networkingv1beta1.VirtualService{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      application.Name + "-ingress",
+			Namespace: application.Namespace,
+		},
+		Spec: networkingv1beta1api.VirtualService{
+			ExportTo: EXPORT_TO_NAMESPACES,
+			Gateways: r.getGatewaysFromApplication(application),
+			Hosts:    application.Spec.Ingresses,
+			Http: []*networkingv1beta1api.HTTPRoute{
+				{
+					Route: []*networkingv1beta1api.HTTPRouteDestination{
+						{
+							Destination: &networkingv1beta1api.Destination{
+								Host: application.Name,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	err := ctrlutil.SetControllerReference(application, &virtualService, r.GetScheme())
+	if err != nil {
+		r.SetControllerError(ctx, application, controllerName, err)
+		return &virtualService, err
+	}
+
+	r.SetLabelsFromApplication(ctx, &virtualService, *application)
+	util.SetCommonAnnotations(&virtualService)
+
+	return &virtualService, err
+}
+
+func (r *ApplicationReconciler) getGatewaysFromApplication(application *skiperatorv1alpha1.Application) []string {
+	gateways := make([]string, 0, len(application.Spec.Ingresses))
+	for _, hostname := range application.Spec.Ingresses {
+		// Generate gateway name
+		hash := fnv.New64()
+		_, _ = hash.Write([]byte(hostname))
+		name := fmt.Sprintf("%s-ingress-%x", application.Name, hash.Sum64())
+		gateways = append(gateways, name)
+	}
+
+	return gateways
 }
+
+func (r *ApplicationReconciler) createOrUpdateVirtualService(ctx context.Context, application skiperatorv1alpha1.Application, virtualService networkingv1beta1.VirtualService) (reconcile.Result, error) {
+	err := r.GetClient().Get(ctx, types.NamespacedName{Namespace: virtualService.Namespace, Name: virtualService.Name}, &virtualService)
+	if errors.IsNotFound(err) {
+		err = r.GetClient().Create(ctx, &virtualService)
+		if err != nil {
+			r.SetControllerError(ctx, &application, controllerName, err)
+			return reconcile.Result{}, err
+		}
+	} else if err != nil {
+		r.SetControllerError(ctx, &application, controllerName, err)
+		return reconcile.Result{}, err
+	} else {
+		err = r.GetClient().Update(ctx, &virtualService)
+		if err != nil {
+			r.SetControllerError(ctx, &application, controllerName, err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, err
+}
+
+var EXPORT_TO_NAMESPACES = []string{".", "istio-system", "istio-gateways"}
```

Comment:
```suggestion
var ExportToNamespaces = []string{".", "istio-system", "istio-gateways"}
```

Use snakeCase and move declaration to the top of the file, under imports.